### PR TITLE
V3RegistrationResource now uses template URL instead of base URL

### PR DESCRIPTION
**Pull request by maartenba**
Note that some resources in the service discovery document at https://api.nuget.org/v3/index.json are rc, some are beta, making the unit tests unhappy. We may want to look at that (outside scope of this PR)

### CODE REVIEW

File: src/NuGet.Client.v3/ResourceSelector.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            foreach (Uri uri in resourceUrls)
+            {
+                if (!cancellationToken.IsCancellationRequested)
+                {
+                    // Get a new HttpClient each time because some BadRequest
+                    // responses were corrupting the HttpClient instance and
+                    // subsequent requests on it would hang unexpectedly
+                    // REVIEW: maballia - would this support proxy / auth scenarios? I guess we need a client that does support those, right?
+                    // REVIEW: maballia - is it ok to always return first available in terms of load balancing?
+                    using (HttpClient http = new HttpClient())
```
Comment (emgarten) [inline src/NuGet.Client.v3/ResourceSelector.cs]: This needs to use the V3HttpHandlerResource when creating a new http client for proxy/auth.

**Comment by emgarten** [comment]
Hitting the server twice for every registration blob is going to slow things down. We need to just fetch the content if it exists instead of doing a head request and coming back to fetch it again right after.

**Comment by maartenba** [comment]
@emgarten That approach was proposed by @jeffhandley in the latest commits of https://github.com/NuGet/NuGet.Protocol/commits/jeffhand/uritemplate - when multiple resources are exposed this one will use the first one to respond.

File: src/NuGet.Client.v3/ResourceSelector.cs (hunk 1 above)
Comment (maartenba) [inline src/NuGet.Client.v3/ResourceSelector.cs]: Fixed with https://github.com/NuGet/NuGet.Protocol/commit/f9429193776955467e37db7e3ba8bf780de8ec95

**Comment by emgarten** [comment]
@maartenba We need to determine the registration blob endpoint up front and stick it until a request timeout is hit. The dependency resolver and UI fetch a large number of registration blobs, any slow down in this path is going to make both of them dramatically slower.

This also needs to handle the scenario where a package does not exist on the server. In the current code it will loop through every possible url in serial, getting 404s but still going on until it has tried everything.

**Comment by maartenba** [comment]
@emgarten So in terms of errors, when should it try the next URL? On 500's? Timeouts?

**Comment by emgarten** [comment]
Both of those sound like a good way to check if the endpoint is down, that would let 404 become a valid  response instead of an error/retry.

**Pull request merged**